signature-envelope-jws.md

```diff
@@ -0,0 +1,165 @@
+}
+```
+
+- **[`x5c`](https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.6)** (*array of strings*): This REQUIRED header contains the ordered list of X.509 certificate or certificate chain([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign the JWS. The certificate chain is represented as a JSON array of certificate value strings, each string in the array is a base64-encoded DER certificate value. The certificate containing the public key corresponding to the key used to digitally sign the JWS MUST be the first certificate, followed by the intermediate and root certificates in the correct order. Refer [*Certificate Chain* unsigned attribute](signature-specification.md#unsigned-attributes) for more details.
+- **`io.cncf.notary.timestamp`** (*string*): This OPTIONAL header is used to store countersignature that provides trusted signing time. Only [RFC3161]([rfc3161](https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2)) compliant `TimeStampToken` are supported.
+  - **TODO** Define the opaque datum (hash of envelope) that is sent to TSA, and how TSA response (time stamp token) is represented in this header.
+- **`io.cncf.notary.signingAgent`**(*string*): This OPTIONAL header provides the identifier of a client (e.g. Notation) that produced the signature. E.g. “notation/1.0.0”. Refer [*Signing Agent* unsigned attribute](signature-specification.md#unsigned-attributes) for more details.
+
+## Signature
+
+In JWS signature is calculated by combining JWSPayload and protected headers.
+The process is described below:
+
+### Create the *JWS Signing Input*
+
+1. Compute the Base64Url value of ProtectedHeaders, this is the value of `protected` property in the signature envelope.
+1. Compute the Base64Url value of JWSPayload, this is the value of `payload` property in the signature envelope.
+1. Build *JWS Signing Input* to be signed by concatenating the values generated in step 1 and step 2 using '.'
+`ASCII(BASE64URL(UTF8(ProtectedHeaders)) ‘.’ BASE64URL(JWSPayload))`
+
+### Generate the signature
+
+1. Compute the signature on the *JWS Signing Input* constructed in the previous step by using the signature algorithm of the signing key, which MUST match the corresponding protected header `alg`.
+2. Compute the Base64Url value of the signature produced in the previous step.
+   This is the value of the `signature` property in the signature envelope.
+
+## Signature Envelope
+
+The final signature envelope comprises of Payload, ProtectedHeaders, UnprotectedHeaders, and Signature, no additional top level fields are supported.
+
+Since Notary v2 restricts one signature per signature envelope, the compliant signature envelope MUST be in flattened JWS JSON format.
+
+```jsonc
+{
+    "payload": "<Base64Url(JWSPayload)>",
+    "protected": "<Base64Url(ProtectedHeaders)>",
+    "header": {
+        "io.cncf.notary.timestamp": "<Base64(TimeStampToken)>",
```

Comment:
```suggestion
        "io.cncf.notary.timestamp": "<Base64(DER(TimeStampToken))>",
```

Reply:
You may be right, I was looking at RFC 3161 which only says DER, not BER. Let's postpone any timestamp-related changes to a follow-up PR.

Reply:
[*Time-Stamp Protocol via HTTP*](https://datatracker.ietf.org/doc/html/rfc3161#section-3.4) in RFC 3161 specifies that the TSA response is ASN.1 DER-encoded. Validated with a [freetsa.org](https://www.freetsa.org) example [here](https://lapo.it/asn1js/#MIIVcjADAgEAMIIVaQYJKoZIhvcNAQcCoIIVWjCCFVYCAQMxDzANBglghkgBZQMEAgMFADCCAaQGCyqGSIb3DQEJEAEEoIIBkwSCAY8wggGLAgEBBgQqAwQBMFEwDQYJYIZIAWUDBAIDBQAEQAy_TK7zgEe7qaJOYhqWFITl0qkhdqhZ5-sn3zQ90065jVOKbF9Noc4wLsJQuCHMAB5GzJenBJiClxhaTffplgICBAHVxjwYDzIwMjIwNTI3MjMwNzU2WgEB_6CCARGkggENMIIBCTERMA8GA1UEChMIRnJlZSBUU0ExDDAKBgNVBAsTA1RTQTF2MHQGA1UEDRNtVGhpcyBjZXJ0aWZpY2F0ZSBkaWdpdGFsbHkgc2lnbnMgZG9jdW1lbnRzIGFuZCB0aW1lIHN0YW1wIHJlcXVlc3RzIG1hZGUgdXNpbmcgdGhlIGZyZWV0c2Eub3JnIG9ubGluZSBzZXJ2aWNlczEYMBYGA1UEAxMPd3d3LmZyZWV0c2Eub3JnMSIwIAYJKoZIhvcNAQkBFhNidXNpbGV6YXNAZ21haWwuY29tMRIwEAYDVQQHEwlXdWVyemJ1cmcxCzAJBgNVBAYTAkRFMQ8wDQYDVQQIEwZCYXllcm6gghAIMIIIATCCBemgAwIBAgIJAMHphhYNqOmCMA0GCSqGSIb3DQEBDQUAMIGVMREwDwYDVQQKEwhGcmVlIFRTQTEQMA4GA1UECxMHUm9vdCBDQTEYMBYGA1UEAxMPd3d3LmZyZWV0c2Eub3JnMSIwIAYJKoZIhvcNAQkBFhNidXNpbGV6YXNAZ21haWwuY29tMRIwEAYDVQQHEwlXdWVyemJ1cmcxDzANBgNVBAgTBkJheWVybjELMAkGA1UEBhMCREUwHhcNMTYwMzEzMDE1NzM5WhcNMjYwMzExMDE1NzM5WjCCAQkxETAPBgNVBAoTCEZyZWUgVFNBMQwwCgYDVQQLEwNUU0ExdjB0BgNVBA0TbVRoaXMgY2VydGlmaWNhdGUgZGlnaXRhbGx5IHNpZ25zIGRvY3VtZW50cyBhbmQgdGltZSBzdGFtcCByZXF1ZXN0cyBtYWRlIHVzaW5nIHRoZSBmcmVldHNhLm9yZyBvbmxpbmUgc2VydmljZXMxGDAWBgNVBAMTD3d3dy5mcmVldHNhLm9yZzEiMCAGCSqGSIb3DQEJARYTYnVzaWxlemFzQGdtYWlsLmNvbTESMBAGA1UEBxMJV3VlcnpidXJnMQswCQYDVQQGEwJERTEPMA0GA1UECBMGQmF5ZXJuMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEAtZEEjE5IbzTp3Ahif8I3UWIjaYS4LLEwvv9RfPw4-EvOXGWodNqyYhrgvOfjNWPg7ek0_V-IIxWfB4SICCJ0YMHtiCYXBvQoEzQ1nfu4G9E1P8F5YQrxqMjIZdwA6iOzqJvmvQO6hansgn1gVlkF4i1qWE7ROArhUCgM7jl-mKAS84BGQAeGJEO8B3y5X0Ia8xcS2Wg8223_uvPIululZq5SPUWdYXc0bU2EDieIa3wBxbiQ14ouJ7uo3S-aKBLhV9YvkhxlliVIBp3Nt9Bt4YHeDpVw1m-HIgzii2KKtVkG8-4MIQ9wUej0hYr4uaktCeRq8tnLpb_PrRaM32BEkaSwZgOxFMr3Ax8GXn7u-lPFdfNJDAWdLjLdx2rE1MTHEGg7l_0b5ZG8YQVRhtiPmgORswe2-R7ZVNqjb5rNah4Uqi5K3xdGS1TbGNu2_-MAgCRlRzcENs5Od7rl3m_g8_nW5_--tGHnlOkvsJUfiq5hpBLM6bIQdGNci-MnrhoPa0pkbrD4RjvGO_hFUwQ10Z6AJRHsn2bDSWlS2L7LabCqTUxB9gUV_n3LuJMZzdpZumrqS-POrnGOb8tszX25_FC7FbEvNmWwqjByicLm3UsRHOSLotnv21prmlBgaTNPs09vx64zDws0IIqsgN8yZv3ZBGWHa6LLiY2VBTFbbnsCAwEAAaOCAdswggHXMAkGA1UdEwQCMAAwHQYDVR0OBBYEFG52C3tOT5zhYMptLOknoqKUs3c3MB8GA1UdIwQYMBaAFPpVDYw0ZlFDTPfns6dsla965qSXMAsGA1UdDwQEAwIGwDAWBgNVHSUBAf8EDDAKBggrBgEFBQcDCDBjBggrBgEFBQcBAQRXMFUwKgYIKwYBBQUHMAKGHmh0dHA6Ly93d3cuZnJlZXRzYS5vcmcvdHNhLmNydDAnBggrBgEFBQcwAYYbaHR0cDovL3d3dy5mcmVldHNhLm9yZzoyNTYwMDcGA1UdHwQwMC4wLKAqoCiGJmh0dHA6Ly93d3cuZnJlZXRzYS5vcmcvY3JsL3Jvb3RfY2EuY3JsMIHGBgNVHSAEgb4wgbswgbgGAQAwgbIwMwYIKwYBBQUHAgEWJ2h0dHA6Ly93d3cuZnJlZXRzYS5vcmcvZnJlZXRzYV9jcHMuaHRtbDAyBggrBgEFBQcCARYmaHR0cDovL3d3dy5mcmVldHNhLm9yZy9mcmVldHNhX2Nwcy5wZGYwRwYIKwYBBQUHAgIwOxo5RnJlZVRTQSB0cnVzdGVkIHRpbWVzdGFtcGluZyBTb2Z0d2FyZSBhcyBhIFNlcnZpY2UgKFNhYVMpMA0GCSqGSIb3DQEBDQUAA4ICAQClyUTixvrAoU2TCn_QoLFytB_BSDw-lXxoorzZuXZPGpUBYf1yRy1Bpe7Sd3hiA7VCIkD7OibN4XYIe2-xAR30zBniVxqkoFEQlmXpTEb1C9Kt7mrEE34lGyWjnavaRRUV2P-eByCejsILeHT34aDt58AJN_6EozT4syZc7S2O2d9hOWWDZ3_rOCwe47I-bqXwXfMN57n4kAXSUmb2EvOci09tq6bXv7rBljK5Bjcyn1Km8GahDkPqqB-Emmxf4_6LXqIydfaH8gUuUC6mwwdipmjM4Hhx3Y6X4xW7qSniVYmXegoxLOlsUQaxQ3x3nys2GxgoiPPuiiNDdPoGPpVhkmJ_fEMQc5ZdEmCSjroAnoA0Ka4yTPlvBCNU83vKWv3cefeTRqs4i_x58B3JhhJU6mzBKZQQdrg9IFVvO-UTJoN_KHb3gzs3Dnw9QQUjgn1PU0AMciGNdSKf8QxviJOpo6HAxCu0yJjBPfQcf2VztPxWUVlxphCnsNKFfIIlqfsgTqzsouiXGqGvh4hqKuPHL-CgquhCmAp3vvFrkhFUWAkNmCtZRmA3ZOdaCtPRFFS5mG9ni5q2r-hJcDOuOr_U60O3vJ3uaIFZSeZIFYKoLnhSd_IoIQfv45AgDgUIrLjqguolBSdvPJ2io9O0rTi7-IQr2jb8JEgpH1WNwC3R4DCCB_8wggXnoAMCAQICCQDB6YYWDajpgDANBgkqhkiG9w0BAQ0FADCBlTERMA8GA1UEChMIRnJlZSBUU0ExEDAOBgNVBAsTB1Jvb3QgQ0ExGDAWBgNVBAMTD3d3dy5mcmVldHNhLm9yZzEiMCAGCSqGSIb3DQEJARYTYnVzaWxlemFzQGdtYWlsLmNvbTESMBAGA1UEBxMJV3VlcnpidXJnMQ8wDQYDVQQIEwZCYXllcm4xCzAJBgNVBAYTAkRFMB4XDTE2MDMxMzAxNTIxM1oXDTQxMDMwNzAxNTIxM1owgZUxETAPBgNVBAoTCEZyZWUgVFNBMRAwDgYDVQQLEwdSb290IENBMRgwFgYDVQQDEw93d3cuZnJlZXRzYS5vcmcxIjAgBgkqhkiG9w0BCQEWE2J1c2lsZXphc0BnbWFpbC5jb20xEjAQBgNVBAcTCVd1ZXJ6YnVyZzEPMA0GA1UECBMGQmF5ZXJuMQswCQYDVQQGEwJERTCCAiIwDQYJKoZIhvcNAQEBBQADggIPADCCAgoCggIBALYCjg4wMvERENlkzalLnQJ44ZQq6ROqpZkHzaaXk5lb2ax-M7rZ_jcE2hwBqY0hr-P1kaWdcGdwUWeZj1AWci4KtGKyH0ORcdLPzEWT83Na95SlqzEfbAEMeJjeM9dcRRDudvS9HRSYzxfTA_BqXdn3lsxsqbZXpW_j6k_vvnzmtqGNPjWjDO5f8XDRzzmjM9P9qJZNIttoWynlYb6JDwqoRYc7LoSrJquDn_6PrenSO7MeYdJzzJuIBkkYX6vs-gU0YAq6kBthTi6FRYLeoiJvwZzX31K-1Q2Hd82ZiMBTo_x9wyh6BopP8StxPNmANmbpVThUVv84-AKYz2uThW6SJHdKZs8c3RHC-O_YUgPXRYslZksT7WOc3tT_gRPWzFNT0nKUc8PDBxV8ciqltd0L-y1sOLG5N0nIgexgAm0IlRs4JL1xusvORzrr1jbwuRi0osj_RpTwdFevLW8c-CVU0XcP15_10xTc0QTN3KvJQTgFbfzwF-frhXL9UvcBRPGI2gX1gj9Y3QYpfnOHvtLXcsE9qCZmAQRf5BLdcJhsDJh7pzRLkDc4dRbSWOeIW1H4lot_JgEhO8TLTIX4_wuEr2qYgzfN-4GGj37PMdymcW1-wt2ALBZyYp5cAFLLNX3Smq_EP2FbOx_51OHOCMccc-H-u33FajNiEynp7WwjAgMBAAGjggJOMIICSjAMBgNVHRMEBTADAQH_MA4GA1UdDwEB_wQEAwIBxjAdBgNVHQ4EFgQU-lUNjDRmUUNM9-ezp2yVr3rmpJcwgcoGA1UdIwSBwjCBv4AU-lUNjDRmUUNM9-ezp2yVr3rmpJehgZukgZgwgZUxETAPBgNVBAoTCEZyZWUgVFNBMRAwDgYDVQQLEwdSb290IENBMRgwFgYDVQQDEw93d3cuZnJlZXRzYS5vcmcxIjAgBgkqhkiG9w0BCQEWE2J1c2lsZXphc0BnbWFpbC5jb20xEjAQBgNVBAcTCVd1ZXJ6YnVyZzEPMA0GA1UECBMGQmF5ZXJuMQswCQYDVQQGEwJERYIJAMHphhYNqOmAMDMGA1UdHwQsMCowKKAmoCSGImh0dHA6Ly93d3cuZnJlZXRzYS5vcmcvcm9vdF9jYS5jcmwwgc8GA1UdIASBxzCBxDCBwQYKKwYBBAGB8iQBATCBsjAzBggrBgEFBQcCARYnaHR0cDovL3d3dy5mcmVldHNhLm9yZy9mcmVldHNhX2Nwcy5odG1sMDIGCCsGAQUFBwIBFiZodHRwOi8vd3d3LmZyZWV0c2Eub3JnL2ZyZWV0c2FfY3BzLnBkZjBHBggrBgEFBQcCAjA7GjlGcmVlVFNBIHRydXN0ZWQgdGltZXN0YW1waW5nIFNvZnR3YXJlIGFzIGEgU2VydmljZSAoU2FhUykwNwYIKwYBBQUHAQEEKzApMCcGCCsGAQUFBzABhhtodHRwOi8vd3d3LmZyZWV0c2Eub3JnOjI1NjAwDQYJKoZIhvcNAQENBQADggIBAGivfr-ThWLvTOs7WAvi-vbMNaJncpYvPZWQH6VjDIfQkZiYTOigajP4qcKC7Z8csRrGwj4XEI7k785vspTelcEzJiJVclUiymGXHUo7f3glDfuNSu7A-xlZsWQQBSC5wQ5kxiZi5K1NCrriKY_JSPxOmejZ5rj9vkQEEh7HwUIurLLJ1zKOBzluYLTzu4A61KVVyA_vtT-F53ZKCp-0r8OZ9M0vX79YcQXGCBzz0FM3trt9GwELdJ9IiMkS82lrobaQLXe338BGwEoMwexPjRheLaVd-3vCogNsYhkkak-Z3btvH4KTmPO4A9wK2Q3LWb70wnx3QEuZBDt4JxhnmRFSw5nxLL_ExiWtwJY1WuRONCEA7FF6UC4vBvlAuNQ1mbvBFU-K52GgsNVV-0oTkdTzQgr42_EvLX3bnXfc4VN4BAdK8XXk8tbVWzS11vfcvdMXMK9WSA1MDP8UP56DvBUYZtC6Dwu9xH_ieGQXa71sGrhd8yXt93eIm8RHG_P6c-VsxZHosWDNp7B4ah7ASsOyT6LijV0Z5eSABNXhZqg8guxv1U-zheuvcTOoW1LeRttSROHDSujTbnEvn84NST19Pt1YbGGY4-w-bpY0b0F6yfIh4K_zOo9qCx70wCNjC3atqo2RQzgl7MQcSaW5ixgcfaMOmXq5VMc8LNgFr9qZMYIDijCCA4YCAQEwgaMwgZUxETAPBgNVBAoTCEZyZWUgVFNBMRAwDgYDVQQLEwdSb290IENBMRgwFgYDVQQDEw93d3cuZnJlZXRzYS5vcmcxIjAgBgkqhkiG9w0BCQEWE2J1c2lsZXphc0BnbWFpbC5jb20xEjAQBgNVBAcTCVd1ZXJ6YnVyZzEPMA0GA1UECBMGQmF5ZXJuMQswCQYDVQQGEwJERQIJAMHphhYNqOmCMA0GCWCGSAFlAwQCAwUAoIG4MBoGCSqGSIb3DQEJAzENBgsqhkiG9w0BCRABBDAcBgkqhkiG9w0BCQUxDxcNMjIwNTI3MjMwNzU2WjArBgsqhkiG9w0BCRACDDEcMBowGDAWBBSRbaPYYOzKguNLxZ0Xk-fpaIdfFDBPBgkqhkiG9w0BCQQxQgRAO-_WBAh7yMfQPSxDF5avIcZjm32iWF6Z4TYhmVof2CfqUZjRwJpV9L4jbQQuaG0wz8GOr5QimT8N9bso--piwjANBgkqhkiG9w0BAQEFAASCAgBaD6X7uR_SR6jJgzL7rgfw3ESroojyNVy_9EyfFWvPHQLSS22GvwW_6zwNgAMPpDwa5tB9k-aU4Xg92vHcGpABKMlpE0QTEbUVjcv-CU1YNaSaMdUsqY2rspqFaRkVSYaoKIAi8Arza6tgNNqmL0EasoomYGmIG3zqKhztrzvptiuwVsIssOTUQBEDVxsUshIOn50Z3HfZvaa9St-ndFl7Bj63O_bBNkxTJGYHDjbhWfeB1MK3oKt_j6UoZGVjgwhRs6iOLgn_P0_WuId3UjymYIfmTqfI6HhaevZff-PCiFR-QMidJzltjoiEVSit93PIDcRSZrgVoB_kHxZ82vVP13bKkKeJzos9j0Ly7rCNCIDG7UCrQ-6nVqPqYWpIlFfDdV5x-gYsETb_OkAEs9UzMmYvQ5rCQZnZl4XMLzSCjx-FEuLnVeoyqNVR56TVo5miGsM1j212lPfTc6is4rG54pwkSKvenSV36qeE8aJEkXAAB8-hb65xCNIOatjJc1tzo8h_ksC0YNQ2LodCcREGel0e2GaY4RDtZsVgUqZkLLIEiUzzhm8iJ8DiNIsUrmAjDxyYqzWn6uQJdmoUWr8-S3abHGtm97AVr3xWuAcEoK34ZZ3_cvINhaFxoumJnE94njtpQ_F-l0Sf5uj_TO0_HEPRtGhdLgm4BVD1me9g-w).

So the current content `"io.cncf.notary.timestamp": "<Base64(TimeStampToken)>"` is accurate, we don't specifically require DER, it's whichever format TSA servers respond with. I'm going to defer changes to timestamp related details, we can refine when we do the implementation. From the signature format's perspective it's an opaque token that is separately validated.